internal/syshandlers/syshandler_tokenpool.go

```diff
@@ -0,0 +1,128 @@
+// Copyright © 2021 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syshandlers
+
+import (
+	"context"
+
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+func (sh *systemHandlers) persistTokenPool(ctx context.Context, pool *fftypes.TokenPoolAnnouncement) (valid bool, err error) {
+	// Find a matching operation within this transaction
+	fb := database.OperationQueryFactory.NewFilter(ctx)
+	filter := fb.And(
+		fb.Eq("tx", pool.TX.ID),
+		fb.Eq("type", fftypes.OpTypeTokensAnnouncePool),
+	)
+	operations, _, err := sh.database.GetOperations(ctx, filter)
+	if err != nil {
+		return false, err // retryable
+	}
+
```

Comment:
Would like some input on the transaction validation logic below. For both cases (operation found locally and operation not found locally), I want to be sure the checks are valid and complete enough.